File: sdk/java/src/main/java/ai/stigmer/sdk/PlatformClientAuth.java
package ai.stigmer.sdk;

import ai.stigmer.iam.platformclient.v1.MintUserTokenRequest;
import ai.stigmer.iam.platformclient.v1.MintUserTokenResponse;
import ai.stigmer.iam.platformclient.v1.PlatformClientTokenControllerGrpc;
import ai.stigmer.sdk.gen.ErrorCode;
import ai.stigmer.sdk.gen.StigmerException;
import io.grpc.ManagedChannel;
import io.grpc.ManagedChannelBuilder;
import io.grpc.Status;
import io.grpc.StatusRuntimeException;
import io.grpc.netty.shaded.io.grpc.netty.NettyChannelBuilder;

import java.time.Instant;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

/**
 * PlatformClient token-minting helper for platform builder backends.
 *
 * <p>A minimal, purpose-built client for minting Stigmer-signed user JWTs.
 * It does NOT replace the main {@link StigmerClient} — use that with an API
 * key for resource management.
 *
 * <p>The returned tokens are passed to the React SDK's {@code StigmerProvider}
 * via {@code getAccessToken} to authenticate browser-based API calls.
 *
 * <pre>{@code
 * try (PlatformClientAuth auth = PlatformClientAuth.builder("api.stigmer.ai:443")
 *         .clientId(System.getenv("STIGMER_CLIENT_ID"))
 *         .clientSecret(System.getenv("STIGMER_CLIENT_SECRET"))
 *         .build()) {
 *     MintUserTokenResult result = auth.mintUserToken(MintUserTokenInput.builder()
 *         .userId("user-123")
 *         .userEmail("jane@acme.com")
 *         .userName("Jane Doe")
 *         .build());
 *     System.out.println(result.accessToken());
 * }
 * }</pre>
 */
public final class PlatformClientAuth implements AutoCloseable {

    private static final long SHUTDOWN_TIMEOUT_SECONDS = 5;

    private final PlatformClientTokenControllerGrpc.PlatformClientTokenControllerBlockingStub tokenStub;
    private final String clientId;
    private final String clientSecret;
    private final ManagedChannel channel;

    private PlatformClientAuth(ManagedChannel channel, String clientId, String clientSecret) {
        this.channel = channel;
        this.clientId = clientId;
        this.clientSecret = clientSecret;
        this.tokenStub = PlatformClientTokenControllerGrpc.newBlockingStub(channel);
    }

    /** Creates a new builder with the given Stigmer API target address. */
    public static Builder builder(String baseUrl) {
        return new Builder(baseUrl);
    }

    /**
     * Mints a user-scoped JWT for browser-based access to Stigmer resources.
     *
     * @param input the user identity and optional org scope
     * @return the minted token with metadata
     * @throws StigmerException with {@link ErrorCode#UNAUTHENTICATED} if credentials are invalid
     * @throws StigmerException with {@link ErrorCode#NOT_FOUND} if the user doesn't exist and
     *         JIT provisioning is disabled
     * @throws StigmerException with {@link ErrorCode#FAILED_PRECONDITION} if the secret has expired
     * @throws StigmerException with {@link ErrorCode#PERMISSION_DENIED} if the origin is not allowed
     */
    public MintUserTokenResult mintUserToken(MintUserTokenInput input) {
        String userId = input.userId();
        if (userId == null || userId.isEmpty()) {
            throw new StigmerException(
                    ErrorCode.INVALID_ARGUMENT,
                    "mintUserToken: userId is required — this is the platform's stable identifier for the user",
                    Status.Code.INVALID_ARGUMENT);
        }

        MintUserTokenRequest request = MintUserTokenRequest.newBuilder()
                .setClientId(clientId)
                .setClientSecret(clientSecret)
                .setUserId(userId)
                .setUserEmail(input.userEmail())
                .setUserName(input.userName())
                .setOrgId(input.orgId())
                .build();

        try {
            MintUserTokenResponse response = tokenStub.mintUserToken(request);
            return new MintUserTokenResult(
                    response.getAccessToken(),
                    response.getTokenType(),
                    response.getExpiresIn(),
                    Instant.now().plusSeconds(response.getExpiresIn()));
        } catch (StatusRuntimeException e) {
            throw StigmerException.wrap(e);
        }
    }

    @Override
    public void close() {
        if (channel.isShutdown()) {
            return;
        }
        channel.shutdown();
        try {
            if (!channel.awaitTermination(SHUTDOWN_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
                channel.shutdownNow();
            }
        } catch (InterruptedException e) {
            channel.shutdownNow();
            Thread.currentThread().interrupt();
        }
    }

    /** Builder for {@link PlatformClientAuth}. */
    public static final class Builder {
        private final String baseUrl;
        private String clientId;
        private String clientSecret;
        private boolean insecure;

        private Builder(String baseUrl) {
            this.baseUrl = Objects.requireNonNull(baseUrl, "baseUrl must not be null");
        }

        /** Sets the PlatformClient client_id (stgm_cid_ prefix). */
        public Builder clientId(String clientId) {
            this.clientId = clientId;
            return this;
        }

        /** Sets the PlatformClient client_secret (stgm_cs_ prefix). Server-only. */
        public Builder clientSecret(String clientSecret) {
            this.clientSecret = clientSecret;
            return this;
        }

        /** Disables TLS. Use only for local development. */
        public Builder insecure() {
            this.insecure = true;
            return this;
        }

        /**
         * Builds the {@link PlatformClientAuth}.
         *
         * @throws IllegalArgumentException if clientId or clientSecret is missing or empty
         */
        public PlatformClientAuth build() {
            if (clientId == null || clientId.isEmpty()) {
                throw new IllegalArgumentException(
                        "stigmer: clientId is required — find it in the Stigmer Console under IAM > Platform Clients");
            }
            if (clientSecret == null || clientSecret.isEmpty()) {
                throw new IllegalArgumentException(
                        "stigmer: clientSecret is required — the secret is shown once at creation time. "
                                + "If lost, rotate via the Console or CLI");
            }

            ManagedChannelBuilder<?> channelBuilder;
            if (insecure) {
                channelBuilder = ManagedChannelBuilder.forTarget(baseUrl).usePlaintext();
            } else {
                channelBuilder = NettyChannelBuilder.forTarget(baseUrl).useTransportSecurity();
            }

            ManagedChannel channel = channelBuilder.build();
            return new PlatformClientAuth(channel, clientId, clientSecret);
        }
    }
}

File: sdk/python/src/stigmer/platform_client_auth.py
"""PlatformClient token-minting helper for platform builder backends.

This module provides a minimal, purpose-built client for minting
Stigmer-signed user JWTs. It is NOT a general-purpose Stigmer client —
use :class:`StigmerClient` with an API key for resource management.

Usage::

    from stigmer.platform_client_auth import platform_client_auth

    auth = platform_client_auth(
        base_url="api.stigmer.ai:443",
        client_id=os.environ["STIGMER_CLIENT_ID"],
        client_secret=os.environ["STIGMER_CLIENT_SECRET"],
    )

    result = auth.mint_user_token(MintUserTokenInput(
        user_id="user-123",
        user_email="jane@acme.com",
        user_name="Jane Doe",
    ))
    print(result.access_token)
"""

from __future__ import annotations

from dataclasses import dataclass
from datetime import datetime, timedelta, timezone
from types import TracebackType

import grpc

from ai.stigmer.iam.platformclient.v1 import io_pb2
from ai.stigmer.iam.platformclient.v1 import token_pb2_grpc

from ._gen._errors import ErrorCode, StigmerError, wrap_error

DEFAULT_TARGET = "api.stigmer.ai:443"


@dataclass(frozen=True)
class MintUserTokenInput:
    """Input for minting a user-scoped Stigmer JWT.

    The platform builder's backend calls this with the authenticated
    user's identity. Stigmer validates the PlatformClient credentials,
    optionally JIT-provisions the user's identity account, and returns
    a signed JWT.
    """

    user_id: str
    """Platform's stable user identifier. Becomes the JWT sub claim."""

    user_email: str = ""
    """User's email address. Used for profile enrichment during JIT provisioning."""

    user_name: str = ""
    """User's display name. Used for profile enrichment during JIT provisioning."""

    org_id: str = ""
    """Organization to scope the token to. Defaults to the PlatformClient's owning org."""


@dataclass(frozen=True)
class MintUserTokenResult:
    """Result of a successful :meth:`PlatformClientAuth.mint_user_token` call.

    Pass ``access_token`` to the React SDK's ``StigmerProvider`` via the
    ``getAccessToken`` callback to authenticate browser-based API calls.
    """

    access_token: str
    """Stigmer-signed JWT for browser-based API authentication."""

    token_type: str
    """Token type. Always "Bearer"."""

    expires_in: int
    """Token lifetime in seconds from issuance."""

    expires_at: datetime
    """Absolute expiration time (UTC), computed from ``expires_in`` at call time."""


class PlatformClientAuth:
    """Mints Stigmer-signed user JWTs from a platform builder's backend.

    Use :func:`platform_client_auth` to create instances — it validates
    configuration before connecting.

    Implements the context manager protocol for clean channel shutdown::

        with platform_client_auth(...) as auth:
            result = auth.mint_user_token(...)
    """

    def __init__(
        self,
        base_url: str,
        client_id: str,
        client_secret: str,
        *,
        insecure: bool = False,
    ) -> None:
        self._client_id = client_id
        self._client_secret = client_secret

        if insecure:
            self._channel = grpc.insecure_channel(base_url)
        else:
            self._channel = grpc.secure_channel(
                base_url, grpc.ssl_channel_credentials()
            )

        self._token_stub = token_pb2_grpc.PlatformClientTokenControllerStub(
            self._channel
        )

    def mint_user_token(self, input: MintUserTokenInput) -> MintUserTokenResult:
        """Mint a user-scoped JWT for browser-based access to Stigmer resources.

        Args:
            input: User identity and optional org scope.

        Returns:
            The minted token with metadata.

        Raises:
            StigmerError: with code ``UNAUTHENTICATED`` if credentials are invalid.
            StigmerError: with code ``NOT_FOUND`` if the user doesn't exist and
                JIT provisioning is disabled.
            StigmerError: with code ``FAILED_PRECONDITION`` if the secret has expired.
            StigmerError: with code ``PERMISSION_DENIED`` if the origin is not allowed.
        """
        if not input.user_id:
            raise StigmerError(
                ErrorCode.INVALID_ARGUMENT,
                "mint_user_token: user_id is required — this is the platform's "
                "stable identifier for the user",
                grpc.StatusCode.INVALID_ARGUMENT,
            )

        request = io_pb2.MintUserTokenRequest(
            client_id=self._client_id,
            client_secret=self._client_secret,
            user_id=input.user_id,
            user_email=input.user_email,
            user_name=input.user_name,
            org_id=input.org_id,
        )

        try:
            response = self._token_stub.mintUserToken(request)
        except grpc.RpcError as e:
            raise wrap_error(e) from e

        return MintUserTokenResult(
            access_token=response.access_token,
            token_type=response.token_type,
            expires_in=response.expires_in,
            expires_at=datetime.now(timezone.utc).replace(microsecond=0)
            + timedelta(seconds=response.expires_in),
        )

    def close(self) -> None:
        """Release the underlying gRPC channel."""
        if self._channel is not None:
            self._channel.close()

    def __enter__(self) -> PlatformClientAuth:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        self.close()


def platform_client_auth(
    base_url: str = DEFAULT_TARGET,
    *,
    client_id: str,
    client_secret: str,
    insecure: bool = False,
) -> PlatformClientAuth:
    """Create a PlatformClient token-minting helper.

    This is the recommended way to mint Stigmer user JWTs from a Python
    backend. The returned tokens are passed to the React SDK's
    ``StigmerProvider`` via ``getAccessToken``.

    Args:
        base_url: gRPC target address (``host:port``). Defaults to
            ``api.stigmer.ai:443``.
        client_id: PlatformClient client_id (``stgm_cid_`` prefix).
        client_secret: PlatformClient client_secret (``stgm_cs_`` prefix).
            Server-only — never expose in browser code.
        insecure: Disable TLS. Use only for local development.

    Returns:
        A :class:`PlatformClientAuth` instance.

    Raises:
        ValueError: If ``client_id`` or ``client_secret`` is empty.
    """
    if not client_id:
        raise ValueError(
            "platform_client_auth: client_id is required — find it in the "
            "Stigmer Console under IAM > Platform Clients"
        )
    if not client_secret:
        raise ValueError(
            "platform_client_auth: client_secret is required — the secret is "
            "shown once at creation time. If lost, rotate via the Console or CLI"
        )

    return PlatformClientAuth(
        base_url=base_url,
        client_id=client_id,
        client_secret=client_secret,
        insecure=insecure,
    )
